### PR TITLE
ref(traffic_simulator, behavior_tree_plugin): hdmap_utils refactor - create lanelet_wrapper, use ::lanelet_map and ::pose

### PR DESCRIPTION
**Pull request by dmoszynski**
# Description

> Please check examples and comment out this sentence. Minimal example is [here](pull_request_samples/example_simple.md) and detailed example is [here](pull_request_samples/example_detail.md)

## Abstract

> [Required] This section is required, keep it short, clear and to the point.  
> Delete this sentence and explain this pull request shortly.  

## Background

> [Optional] If there is no background information that needs explanation (e.g., just a typo correction, etc.), you can skip this section.  
> Delete this sentence and explain the circumstances that led to this pull request being sent.  

## Details

> [Optional] If there are only differences whose effects are so obvious that no explanation is needed, or if there are no differences in the code (e.g., documentation additions), you can skip this section.  
> Delete this sentence and describe this pull request.  
> For example, it is desirable to describe the specifications of added functions, and detailed explanations of bugs that have been fixed.  

## References

> [Optional] If the referenced material does not exist, you can skip this section.  
> Describe any standards, algorithms, books, articles, etc. that you referenced when creating the pull request. If there is any novelty, mention it.  

# Destructive Changes

> [Optional] If no destructive change exists, you can skip this section.  
> Include a description of the changes and a migration guide and send the pull request with a bump major label. (Example : https://github.com/tier4/scenario_simulator_v2/pull/1139)  
> Otherwise, skip the "Destructive Changes" section and make sure this pull request is labeled `bump minor` or `bump patch`.  

# Known Limitations

> [Optional] If there are no known limitations that you can think of, you can skip this section. If there are any limitations on the features or fixes added by this pull request, delete this sentence and clearly describe them.
> For example, the lane matching algorithm currently (1/25/2024) employed is unable to match Entity that is heavily tilted with respect to the lane, and it is difficult to throw an exception.  
> If the developer is aware of the existence of such problems or limitations, describe them in detail. The problems or limitation should be listed as an Issue on GitHub and a link to it should be provided in this section.

### CODE REVIEW

**Comment by dmoszynski** [comment]
The non-regression is evidenced [here](https://github.com/tier4/sim_evaluation_tools/issues/424#issuecomment-2550543940).

File: simulation/traffic_simulator/src/lanelet_wrapper/pose.cpp [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+      }
+    }
+    if (!found_next_lanelet_in_route) {
+      return {std::nullopt, canonicalized_lanelet_pose.lanelet_id};
+    }
+  }
+  return {canonicalized_lanelet_pose, std::nullopt};
+}
+
+// used only by this namespace
```
Comment (hakuturu583) [inline simulation/traffic_simulator/src/lanelet_wrapper/pose.cpp:389]: Does this mean that the function is not exposed to the outside world?
If so, make it impossible to call it externally by removing the forward declaration from the .hpp file, for example.

**Review by hakuturu583 — CHANGES_REQUESTED** [review]
Lengthy PRs such as this one should be avoided as the risk of degrade is very high.
Even in this case, it would have been possible to split the PR by moving it little by little instead of moving many functions.

File: simulation/behavior_tree_plugin/src/action_node.cpp [hunk 2]
```diff
@@ -301,13 +302,9 @@ auto ActionNode::getDistanceToTargetEntityPolygon(
   double width_extension_right, double width_extension_left, double length_extension_front,
   double length_extension_rear) const -> std::optional<double>
 {
-  const auto & status = getEntityStatus(target_name);
-  if (status.laneMatchingSucceed()) {
```
[the remark this answers is not in the document]
Comment (dmoszynski) [reply simulation/behavior_tree_plugin/src/action_node.cpp:305]: Since the precondition in the basic version of `getDistanceToTargetEntityPolygon` is the result==true of the call `isOtherEntityAtConsideredAltitude(status)` <- `status.laneMatchingSucceed()` check is mainly located there.
The occurrence of the check `status.laneMatchingSucceed()` in other, earlier places is redundant. 
So I removed them and ensured that the check is done only in one place.

https://github.com/tier4/scenario_simulator_v2/blob/52b054cc4c21e1d15db16ea0fbbc8cf070611ac4/simulation/behavior_tree_plugin/src/action_node.cpp#L327-L337


Background:
https://github.com/tier4/scenario_simulator_v2/blob/52b054cc4c21e1d15db16ea0fbbc8cf070611ac4/simulation/behavior_tree_plugin/src/action_node.cpp#L300-L325

File: simulation/behavior_tree_plugin/src/action_node.cpp [hunk 3]
```diff
@@ -316,21 +313,27 @@ auto ActionNode::getDistanceToTargetEntityPolygon(
   double width_extension_left, double length_extension_front, double length_extension_rear) const
   -> std::optional<double>
 {
-  if (status.laneMatchingSucceed() && isOtherEntityAtConsideredAltitude(status)) {
```
[the remark this answers is not in the document]
Comment (dmoszynski) [reply simulation/behavior_tree_plugin/src/action_node.cpp:319]: Please read [this comment.](https://github.com/tier4/scenario_simulator_v2/pull/1472#discussion_r1916320461)

File: simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+class RouteCache
+{
+public:
+  auto getRoute(
+    const lanelet::Id from_lanelet_id, const lanelet::Id to_lanelet_id,
+    const lanelet::LaneletMapPtr & lanelet_map, const RoutingConfiguration & routing_configuration,
+    const lanelet::routing::RoutingGraphConstPtr & routing_graph) -> lanelet::Ids
+  {
+    if (!exists(from_lanelet_id, to_lanelet_id, routing_configuration.allow_lane_change)) {
+      constexpr int routing_cost_id = 0;
```
[the remark this answers is not in the document]
Comment (dmoszynski) [reply simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp:79]: I'm not quite sure why the value ‘0’ is set - below previous version:

https://github.com/tier4/scenario_simulator_v2/blob/52b054cc4c21e1d15db16ea0fbbc8cf070611ac4/simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp#L1766-L1768

But as far as I could find, it allows us to use the default route cost - as the distance traveled along the lanelets.
Is this correct and is the following comment correct?

`constexpr int routing_cost_id = 0; ///< to use default routing costs: traveled distance along lanelets`

File: simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+      const auto p2 = center_points[1];
+      const auto p1 = geometry_msgs::build<Point>()
+                        .x((p0.x + p2.x) * 0.5)
+                        .y((p0.y + p2.y) * 0.5)
+                        .z((p0.z + p2.z) * 0.5);
+      center_points.clear();
+      center_points.push_back(p0);
+      center_points.push_back(p1);
+      center_points.push_back(p2);
+    }
```
[the remark this answers is not in the document]
Comment (dmoszynski) [reply simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp:210]: To be honest, I don't really know the answer to this question.
I simply kept the previous implementation as it was - if the number of points is equal to 2 a point between them is added, you can see it in [here](https://github.com/tier4/scenario_simulator_v2/blob/22f11aa17a26829352309038aef4de4fe51a9a34/simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp#L992):

https://github.com/tier4/scenario_simulator_v2/blob/22f11aa17a26829352309038aef4de4fe51a9a34/simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp#L973-L991

File: simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/traffic_rules.hpp [hunk 6]
```diff
@@ -29,16 +31,17 @@ struct Locations
...
   /// @note this function overrides and adds road shoulder handling to GenericTrafficRules::canPass
-  lanelet::Optional<bool> canPass(
-    const std::string & type, const std::string & /*location*/) const override
+  auto canPass(const std::string & type, const std::string & /*location*/) const
+    -> lanelet::Optional<bool> override
   {
     using lanelet::AttributeValueString;
     using lanelet::Participants;
-    const static std::map<std::string, std::vector<std::string>> participants_map{
+    const static std::map<std::string, std::vector<std::string>, std::less<>> participants_map{
```
[the remark this answers is not in the document]
Comment (dmoszynski) [reply simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/traffic_rules.hpp:44]: In the previous version there was also a `std::map`:
https://github.com/tier4/scenario_simulator_v2/blob/22f11aa17a26829352309038aef4de4fe51a9a34/simulation/traffic_simulator/include/traffic_simulator/hdmap_utils/traffic_rules.hpp#L41
I don't see much difference between using `std::map` and `std::unordered_map`, so I left it as it was.

For more flexibility (in types to be compared) I've added `std::less<>` here - as was suggested by the sonar tool.

If you see the advantages of using `std::unordered_map` here (for such a number of keys) please let me know and I will make this change.

File: simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp [hunk 7]
```diff
@@ -218,18 +102,18 @@ auto HdMapUtils::countLaneChanges(
...
-          lane_changes.first++;
-        } else if (auto rights =
-                     getRightLaneletIds(previous, routing_configuration.routing_graph_type, false);
-                   std::find(rights.begin(), rights.end(), current) != rights.end()) {
-          lane_changes.second++;
-        }
+      if (const auto followings =
+            lanelet_map::nextLaneletIds(previous, routing_configuration.routing_graph_type);
+          std::find(followings.begin(), followings.end(), current) != followings.end()) {
+        continue;
```
[the remark this answers is not in the document]
Comment (dmoszynski) [reply simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp:108]: Could you tell me what change you meen?
Is it about a change from `== followings.end()` to `!= followings.end()` and condition construction?

 or 

Is it about a change from 
`auto followings = getNextLaneletIds(previous, routing_configuration.routing_graph_type)` 
to 
`const auto followings = lanelet_map::nextLaneletIds(previous, routing_configuration.routing_graph_type)`
which was required due to the removal of `HdMapUtils::getNextLaneletIds()`?

master version as background:
https://github.com/tier4/scenario_simulator_v2/blob/22f11aa17a26829352309038aef4de4fe51a9a34/simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp#L222-L223

File: simulation/traffic_simulator/src/utils/lanelet_map.cpp [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+namespace traffic_simulator
+{
+inline namespace lanelet_map
+{
+auto laneletLength(const lanelet::Id lanelet_id) -> double
+{
+  return lanelet_wrapper::lanelet_map::laneletLength(lanelet_id);
+}
+
+auto borderlinePoses() -> std::vector<std::pair<lanelet::Id, Pose>>
```
[the remark this answers is not in the document]
Comment (dmoszynski) [reply simulation/traffic_simulator/src/utils/lanelet_map.cpp:26]: You're right, `borderlinePoses` is not very good here.
How do you feel about `noNextLaneletPoses` / `terminalPoses` / `endLaneletsPoses`? - this function returns a list of poses at the ends of those lanelets that do not have `nextLanelet`

In my opinion the best would be `noNextLaneletPoses`, what do you think?

File: simulation/traffic_simulator/test/src/traffic_lights/test_traffic_lights.cpp [hunk 9]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2024 TIER IV, Inc. All rights reserved.
+// Copyright 2015 TIER IV, Inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (dmoszynski) [reply simulation/traffic_simulator/test/src/traffic_lights/test_traffic_lights.cpp:1]: Please take a look at [this comment](https://github.com/tier4/scenario_simulator_v2/pull/1406#discussion_r1802603347) (and others similar). 
Based on experience from previous PRs, we use the date 2015. 
Please let me know which version is finally right. :bow:

File: simulation/traffic_simulator/test/src/traffic_lights/helper.hpp [hunk 10]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2024 TIER IV, Inc. All rights reserved.
+// Copyright 2015 TIER IV, Inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (dmoszynski) [reply simulation/traffic_simulator/test/src/traffic_lights/helper.hpp:1]: Please [see](https://github.com/tier4/scenario_simulator_v2/pull/1472#discussion_r1916410241)

File: simulation/traffic_simulator/src/lanelet_wrapper/lanelet_loader.cpp [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+  const auto accumulated_lengths = calculateAccumulatedLengths(line_string);
+
+  auto findNearestIndexPair =
+    [&](const double target_length) -> std::pair<std::size_t, std::size_t> {
+    const auto N = accumulated_lengths.size();
+    if (target_length < accumulated_lengths.at(1)) {  // Front
+      return std::make_pair(0, 1);
+    } else if (target_length > accumulated_lengths.at(N - 2)) {  // Back
+      return std::make_pair(N - 2, N - 1);
+    } else  // Middle
```
[the remark this answers is not in the document]
Comment (dmoszynski) [reply simulation/traffic_simulator/src/lanelet_wrapper/lanelet_loader.cpp:101]: Unfortunately I'm not the author of this method or the comments, so i'm not sure what the author had in mind, 
maybe it's worth just delete comments front / back / middle? :bow:

File: simulation/traffic_simulator/src/lanelet_wrapper/pose.cpp [hunk 12]
```diff
@@ -0,0 +1,471 @@
...
+  }
+  return std::nullopt;
+}
+
+auto toLaneletPose(
+  const Pose & map_pose, const bool include_crosswalk, const double matching_distance)
+  -> std::optional<LaneletPose>
+{
+  constexpr double distance_threshold{0.1};
+  constexpr std::size_t search_count{5};
```
[the remark this answers is not in the document]
Comment (dmoszynski) [reply simulation/traffic_simulator/src/lanelet_wrapper/pose.cpp:143]: `distance_threshold` equal to `0.1` and `search_count` equal to `5` were added in the previous version, 
I'm not sure why these values, do you have an idea? :bow: 

Or maybe adding
 `/// There is no technical basis for this value, it was determined based on experiments` is sufficient?

File: simulation/traffic_simulator/src/lanelet_wrapper/pose.cpp [hunk 13]
```diff
@@ -0,0 +1,471 @@
...
+      !lanelet_pose_s) {
+    return std::nullopt;
+  } else if (const auto pose_on_centerline = lanelet_spline->getPose(lanelet_pose_s.value());
+             !isAltitudeMatching(map_pose.position.z, pose_on_centerline.position.z)) {
+    return std::nullopt;
+  } else {
+    if (const auto lanelet_pose_rpy = math::geometry::convertQuaternionToEulerAngle(
+          math::geometry::getRotation(pose_on_centerline.orientation, map_pose.orientation));
+        std::fabs(lanelet_pose_rpy.z) > M_PI * yaw_threshold &&
+        std::fabs(lanelet_pose_rpy.z) < M_PI * (1 - yaw_threshold)) {
```
[the remark this answers is not in the document]
Comment (dmoszynski) [reply simulation/traffic_simulator/src/lanelet_wrapper/pose.cpp]: I've made changes so that the `M_PI` is used earlier to calculate the range: [see this commit](https://github.com/tier4/scenario_simulator_v2/pull/1472/commits/79517b6cf865163076c399cbafdf96c0a8ac20b3).
 I‘m not sure if this is to what you expect, please let me know. :bow:

File: simulation/traffic_simulator/src/lanelet_wrapper/pose.cpp (hunk 1 above)
Comment (dmoszynski) [reply simulation/traffic_simulator/src/lanelet_wrapper/pose.cpp]: Yes, but I couldn't do it at the moment because the old version of HdMapUtils still uses it - I've deleted the comments for now.

File: simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+        for (const auto & lanelet : route->shortestPath()) {
+          shortest_path_ids.push_back(lanelet.id());
+        }
+        appendData(
+          from_lanelet_id, to_lanelet_id, routing_configuration.allow_lane_change,
+          shortest_path_ids);
+      }
+    }
+    std::lock_guard lock(mutex_);
+    return data_.at({from_lanelet_id, to_lanelet_id, routing_configuration.allow_lane_change});
```
[the remark this answers is not in the document]
Comment (TauTheLepton) [reply simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp:98]: Could you elaborate on what exactly adds more processing?

### Code comparison
This function:

https://github.com/tier4/scenario_simulator_v2/blob/380eb7be60f45ee9d3ff89b624954fc87da7b7d0/simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp#L73-L99

Is a slightly refactored copy of this function from `master` branch:

https://github.com/tier4/scenario_simulator_v2/blob/22f11aa17a26829352309038aef4de4fe51a9a34/simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp#L2275-L2304

Aside from the changes that were necessary due to the change of the structure, the logic is identical.

### Analysis
The code you have highlighted:

https://github.com/tier4/scenario_simulator_v2/blob/380eb7be60f45ee9d3ff89b624954fc87da7b7d0/simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp#L97-L98

is not exactly the same as the version on master, but it is analogous to this code:

https://github.com/tier4/scenario_simulator_v2/blob/22f11aa17a26829352309038aef4de4fe51a9a34/simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp#L2282

The only difference is that later uses this function:

https://github.com/tier4/scenario_simulator_v2/blob/22f11aa17a26829352309038aef4de4fe51a9a34/simulation/traffic_simulator/include/traffic_simulator/hdmap_utils/cache.hpp#L57-L68

As you can see, the function `getRoute` does essentially the same, but it also additionally checks whether the route exists.

In the new code the check has already been performed, and when the route does not exist it is calculated, so it is guaranteed that at the end of this function the route will already exist.
https://github.com/tier4/scenario_simulator_v2/blob/380eb7be60f45ee9d3ff89b624954fc87da7b7d0/simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp#L78

> [!NOTE]
> Please note that the main condition has been inverted and that is why the order of code is changed.
>
> `master`:
> https://github.com/tier4/scenario_simulator_v2/blob/22f11aa17a26829352309038aef4de4fe51a9a34/simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp#L2281
>
> this PR:
> https://github.com/tier4/scenario_simulator_v2/blob/380eb7be60f45ee9d3ff89b624954fc87da7b7d0/simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp#L78

File: simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (hakuturu583) [reply simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp]: Please leave comment below.

```cpp
/// @noot parameter met harde draad
/// @sa https://docs.ros.org/en/noetic/api/lanelet2_routing/html/doxygen/classlanelet_1_1routing_1_1RoutingGraph.html#a8b11a0cb1541ff2bee670bfbb9fc3a02
constexpr int routing_cost_id = 0; 
```

File: simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (hakuturu583) [reply simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp:98]: [Near here](https://github.com/tier4/scenario_simulator_v2/blob/52b054cc4c21e1d15db16ea0fbbc8cf070611ac4/simulation/traffic_simulator/ include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp#L78-L98), the lock is not allocated even though it is writing to a shared resource called member variable.

It is highly likely that the critical section has changed due to the addition of processing, so please analyze and modify the implementation to be thread-safe.

Also, adding functionality to a large PR like this should not be done, as this kind of problem will occur.
In the future, please do not add features when refactoring unless it is impossible to do so.

File: simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (hakuturu583) [reply simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp:210]: I understand that you have taken over the previous implementation.
Thank you.

File: simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/traffic_rules.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (hakuturu583) [reply simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/traffic_rules.hpp:44]: I understand that you have taken over the previous implementation.
Thank you.
If that is the case, I think it would be better to fix it in a different PR in the future, so I will withdraw this point.

File: simulation/traffic_simulator/src/utils/lanelet_map.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (hakuturu583) [reply simulation/traffic_simulator/src/utils/lanelet_map.cpp:26]: I agree, `noNextLaneletPoses` is the best.

File: simulation/traffic_simulator/src/lanelet_wrapper/pose.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (hakuturu583) [reply simulation/traffic_simulator/src/lanelet_wrapper/pose.cpp:148]: If you do not know what the value means,

```cpp
/// @note Hardcoded parameter, this value has no technical basis and is determined based on experimentation.
/// @todo Add doxygen comments as soon as you know the meaning and rationale of the value.
```

and leave it in the comments.

File: simulation/traffic_simulator/src/lanelet_wrapper/lanelet_loader.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (hakuturu583) [reply simulation/traffic_simulator/src/lanelet_wrapper/lanelet_loader.cpp:101]: Was this statement copied and pasted from a function in the master branch from somewhere?
If so, you can delete it.
If it was added in this PR, please add a comment appropriately.

File: simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (hakuturu583) [reply simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp:107]: I judge the code to have the same behavior as the master branch as far as I review it, but when it comes to the number of lines in this diff, I can't make an accurate judgment.

When submitting a refactor PR like this one, it is preferable to improve the logic in a separate PR after writing todo comments even if improvements can be made to the existing code.

[the remark this answers is not in the document]
Comment (hakuturu583) [reply simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp:107]: This is a comment and a request to improve the efficiency of future Pull Request reviews.
So please resolve conversation after reviewing the text.

File: simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (hakuturu583) [reply simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp]: OK.
But please leave comment in doxygen format.

[the remark this answers is not in the document]
Comment (dmoszynski) [reply simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp]: Please take a look at section [Putting documentation after members](https://www.doxygen.nl/manual/docblocks.html#memberdoc) in the doxygen documentation for C++.
 According to it, my proposal
```
constexpr int routing_cost_id = 0; ///< to use default routing costs: traveled distance along lanelets
```
 is in doxygen format, do you think it's better to not to use `///<` or just `///` alone in the way I proposed?

If you think it is a good idea to use `@note` commonly - from now on we will always add it to every added comment. 
If so, please confirm that if `@note` is used - should the comment always be above the code (not beside it)?

In other words, please let us know which of the following options are acceptable:
### 1
```
constexpr int variable = 0; ///< comment
```
### 2
```
constexpr int variable = 0; /// comment
```
### 3
```
constexpr int variable = 0; ///< @note comment
```
### 4
```
constexpr int variable = 0; /// @note comment
```
### 5
```
/// @note comment
constexpr int variable = 0; 
```

File: simulation/traffic_simulator/src/lanelet_wrapper/lanelet_loader.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (TauTheLepton) [reply simulation/traffic_simulator/src/lanelet_wrapper/lanelet_loader.cpp:101]: I believe, that after some changes (like conversion to a lambda) this is the equivalent code on master branch

https://github.com/tier4/scenario_simulator_v2/blob/22f11aa17a26829352309038aef4de4fe51a9a34/simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp#L1665-L1681

File: simulation/traffic_simulator/src/utils/lanelet_map.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (TauTheLepton) [reply simulation/traffic_simulator/src/utils/lanelet_map.cpp]: I have changed the name of `borderlinePoses` to `noNextLaneletPoses` in [this commit](https://github.com/tier4/scenario_simulator_v2/pull/1472/commits/94a331507cb9b7afb5473bf2b6b9da749485a26c).

File: simulation/traffic_simulator/src/lanelet_wrapper/pose.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (TauTheLepton) [reply simulation/traffic_simulator/src/lanelet_wrapper/pose.cpp:150]: I have added such comment in [this commit](https://github.com/tier4/scenario_simulator_v2/pull/1472/commits/1fca5858b4e1a9cd2fe0c1fc63c21c268d8ccf79).

File: simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (TauTheLepton) [reply simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp]: I am not sure if I fully understand your intention, but based on my understanding, I have prepared a proposal in [this commit](https://github.com/tier4/scenario_simulator_v2/pull/1472/commits/41f9dd1e226ea827444aed1d01c207d10b9b61fa).

My proposition is to secure the whole function body with a lock_guard, see the code below

https://github.com/tier4/scenario_simulator_v2/blob/41f9dd1e226ea827444aed1d01c207d10b9b61fa/simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp#L78-L98

This will prevent any change to the member variable `data_` made by some other thread for the whole time of execution of this function.

This philosophy has been incorporated in the mentioned commit for all functions of cache classes. So the resulting implementation is such that the public member functions lock the mutex for their whole execution, whereas the private member functions are not thread safe on their own and need to be called from the thread safe scope (the public member functions in this case).

File: simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (hakuturu583) [reply simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp]: Adding @note is best.
Please use 5 format.

```cpp
/// @note comment
constexpr int variable = 0; 
```

[the remark this answers is not in the document]
Comment (TauTheLepton) [reply simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp:81]: I have adjusted all comments in this PR to comply with format `\\\ @note` in [this commit](https://github.com/tier4/scenario_simulator_v2/pull/1472/commits/f4c4503d2325e9d884c81763386ac2345a8ce038).

File: simulation/traffic_simulator/src/lanelet_wrapper/lanelet_loader.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (TauTheLepton) [reply simulation/traffic_simulator/src/lanelet_wrapper/lanelet_loader.cpp]: According to your request, I have removed these comments [here](https://github.com/tier4/scenario_simulator_v2/pull/1472/commits/b3eeab579528d5f4440d97f85fa39efa817ab7f7).

File: simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (TauTheLepton) [reply simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp:99]: I have overridden the implementation with the 2nd version, as requested [here](https://tier4.atlassian.net/browse/RJD-1546?focusedCommentId=204764).
I have allowed myself to keep the modification of restricting the access to data members and some member functions as `private`. I believe keeping the whole class as `public` was an oversight.

File: simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp [hunk 15]
```diff
@@ -0,0 +1,306 @@
...
+  }
+
+  auto getLength(const lanelet::Id lanelet_id, const lanelet::LaneletMapPtr & lanelet_map) -> double
+  {
+    if (!exists(lanelet_id)) {
+      appendData(
+        lanelet_id, lanelet::utils::getLaneletLength2d(lanelet_map->laneletLayer.get(lanelet_id)));
+    }
+    std::lock_guard lock(mutex_);
+    return data_.at(lanelet_id);
```
Comment (hakuturu583) [inline simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp:237]: Please add function below.

```cpp
auto readData(const lanelet::Id lanelet_id) -> double
{
    std::lock_guard lock(mutex_);
    return data_.at(lanelet_id);
}
```

and modify getLength to

```cpp
  auto getLength(const lanelet::Id lanelet_id, const lanelet::LaneletMapPtr & lanelet_map) -> double
  {
    if (!exists(lanelet_id)) {
      appendData(
        lanelet_id, lanelet::utils::getLaneletLength2d(lanelet_map->laneletLayer.get(lanelet_id)));
    }
    return readData(lanelet_id);
}
```

Comment (hakuturu583) [reply simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp:237]: Please correct other similar sections in the same way.

**Review by hakuturu583 — CHANGES_REQUESTED** [review]

File: simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp (hunk 15 above)
Comment (TauTheLepton) [reply simulation/traffic_simulator/include/traffic_simulator/lanelet_wrapper/lanelet_wrapper.hpp:237]: Certainly :bowing_man:

**Review by hakuturu583 — APPROVED** [review]
LGTM
